reV/SAM/SAM.py

```diff
@@ -613,37 +578,46 @@ def drop_leap(resource):
         return resource
 
     @staticmethod
-    def ensure_res_len(res_arr, base=8760):
-        """Ensure that the length of resource array is a multiple of base.
+    def ensure_res_len(arr, time_index):
+        """
+        Ensure time_index has a constant time-step and only covers 365 days
+        (no leap days). If not remove last day
 
         Parameters
         ----------
-        res_arr : np.ndarray
-            Array of resource data.
-        base : int
-            Ensure that length of resource array is a multiple of this value.
+        arr : ndarray
+            Array to truncate if time_index has a leap day
+        time_index : pandas.DatatimeIndex
+            Time index associated with arr, used to check time-series
+            frequency and number of days
 
         Returns
         -------
-        res_arr : array-like
-            Truncated array of resource data such that length(res_arr)%base=0.
+        arr : ndarray
+            Truncated array of data such that there are 365 days
         """
-
-        if len(res_arr) < base:
-            msg = ('Received timeseries of length {}, expected timeseries to'
-                   'be at least {}'.format(len(res_arr), base))
-            logger.exception(msg)
+        msg = ('A valid time_index must be supplied to ensure the proper '
+               'resource length! Instead {} was supplied'
+               .format(type(time_index)))
+        assert isinstance(time_index, pd.DatetimeIndex)
+
+        msg = ('arr length {} does not match time_index length {}!'
+               .format(len(arr), len(time_index)))
+        assert len(arr) == len(time_index)
+
+        freq = pd.infer_freq(time_index)
+        if freq is None:
+            msg = ('Resource time_index does not have a consistent time-step '
+                   '(frequency)!')
```

Comment:
This will fail if we mask out feb 29.

Reply:
It doesn't!
In [1]: import pandas as pd

In [2]: test = pd.date_range('2012', '2013', freq='h', closed='left)
  File "<ipython-input-2-a06fbcd997d5>", line 1
    test = pd.date_range('2012', '2013', freq='h', closed='left)
                                                                ^
SyntaxError: EOL while scanning string literal


In [3]: test = pd.date_range('2012', '2013', freq='h', closed='left')

In [4]: test
Out[4]:
DatetimeIndex(['2012-01-01 00:00:00', '2012-01-01 01:00:00',
               '2012-01-01 02:00:00', '2012-01-01 03:00:00',
               '2012-01-01 04:00:00', '2012-01-01 05:00:00',
               '2012-01-01 06:00:00', '2012-01-01 07:00:00',
               '2012-01-01 08:00:00', '2012-01-01 09:00:00',
               ...
               '2012-12-31 14:00:00', '2012-12-31 15:00:00',
               '2012-12-31 16:00:00', '2012-12-31 17:00:00',
               '2012-12-31 18:00:00', '2012-12-31 19:00:00',
               '2012-12-31 20:00:00', '2012-12-31 21:00:00',
               '2012-12-31 22:00:00', '2012-12-31 23:00:00'],
              dtype='datetime64[ns]', length=8784, freq='H')

In [5]: mask = test.month == 29

In [6]: mask &= test.day == 2

In [7]: masked = test[~mask]

In [8]: pd.infer_freq(masked)
Out[8]: 'H'

Reply:
but what if month==2 and day==29?

Reply:
I don't follow? In the above I masked out 2/29...

Reply:
No you didnt :)

Reply:
In [5]: mask = test.month == 29
In [6]: mask &= test.day == 2

How is that not 2/29?

Reply:
Oh shit, duh 29/2

Reply:
No wonder it worked :-)
=

Reply:
Okay, how does the updated logic look? I also found a bug with the doy logic where 12/31 would get dropped even if 2/29 was dropped. Fixed that and added a test for both issues

Reply:
nice yeah that looks great!